### PR TITLE
Local server support Tun interface for both TCP and UDP

### PR DESCRIPTION
**Pull request by zonyitoo**
ref #581

### TCP

image

1. Create a `TcpListener` listening on an address in the subnet of `tun` device.
2. When `tun` receives TCP packets, it will redirect it to `TcpListener` by modifying `saddr` to a unique fake address and `daddr` to the address that `TcpListener` was listening.
3. `TcpListener` accepts a new connection. The connection's `peer_addr` will be the unique fake address. So we can get the original destination address by looking to the `tun`'s address allocation map.

### UDP

The same as the `redir` design, except it sends and receives packets from the `tun` interface.

### Possible Applications

- shadowsocks-android
- shadowsocks-ios
- macOS & Linux
- Windows (with the [Wintun](https://www.wintun.net/) project)

### CODE REVIEW

**Comment by dev4u** [comment]
噢，又有新玩法了！
什么时候可以尝鲜测试？

**Comment by zonyitoo** [comment]
This is already available for trial.

**Comment by zonyitoo** [comment]
> 按照主题的结构图，plugin会在什么位置上呢？

Just like a normal `sslocal`, runs after the `TcpListener`.

**Comment by dev4u** [comment]
> This is already available for trial.

好，我找时间试试。

**Comment by dev4u** [comment]
```log
error[E0061]: this function takes 1 argument but 2 arguments were supplied
   --> bin/sslocal.rs:359:48
    |
359 |                     let mut local_dns_config = LocalConfig::new(addr, ProtocolType::Dns);
    |                                                ^^^^^^^^^^^^^^^^ ----  ----------------- supplied 2 arguments
    |                                                |
    |                                                expected 1 argument
    |
note: associated function defined here
   --> /Users/anders/develop/workspaces/shadowsocks-android/core/src/main/rust/shadowsocks-rust/crates/shadowsocks-service/src/config.rs:693:12
    |
693 |     pub fn new(protocol: ProtocolType) -> LocalConfig {
    |            ^^^
```
放到android里编译失败了。

**Comment by dev4u** [comment]
```log
> Task :core:cargoBuildArm64
   Compiling shadowsocks-service v1.12.0 (/workspaces/shadowsocks-android/core/src/main/rust/shadowsocks-rust/crates/shadowsocks-service)
error[E0432]: unresolved import `tun::AsyncDevice`
  --> crates/shadowsocks-service/src/local/tun/mod.rs:21:11
   |
21 | use tun::{AsyncDevice, Configuration as TunConfiguration, Device, Layer};
   |           ^^^^^^^^^^^ no `AsyncDevice` in the root

error[E0425]: cannot find function `create_as_async` in crate `tun`
  --> crates/shadowsocks-service/src/local/tun/mod.rs:89:33
   |
89 |         let device = match tun::create_as_async(&self.tun_config) {
   |                                 ^^^^^^^^^^^^^^^ not found in `tun`

error[E0599]: no method named `poll` found for struct `Pin<&mut _>` in the current scope
   --> crates/shadowsocks-service/src/local/tun/mod.rs:165:13
    |
165 | /             tokio::select! {
166 | |                 // tun device
167 | |                 n = self.device.read(&mut packet_buffer) => {
168 | |                     let n = n?;
...   |
197 | |                 }
198 | |             }
    | |_____________^ method not found in `Pin<&mut _>`
    |
    = note: `fut` is a function, perhaps you wish to call it
    = note: this error originates in the macro `$crate::select` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `poll` found for struct `Pin<_>` in the current scope
   --> crates/shadowsocks-service/src/local/tun/mod.rs:165:13
    |
165 | /             tokio::select! {
166 | |                 // tun device
167 | |                 n = self.device.read(&mut packet_buffer) => {
168 | |                     let n = n?;
...   |
197 | |                 }
198 | |             }
    | |_____________^ method not found in `Pin<_>`
    |
    = note: `fut` is a function, perhaps you wish to call it
    = note: this error originates in the macro `$crate::select` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no method named `packet_information` found for mutable reference `&mut tun::platform::Configuration` in the current scope
  --> crates/shadowsocks-service/src/local/tun/mod.rs:86:24
   |
86 |             tun_config.packet_information(false);
   |                        ^^^^^^^^^^^^^^^^^^ method not found in `&mut tun::platform::Configuration`

warning: unused import: `AsyncReadExt`
unused import: `AsyncReadExt`

  --> crates/shadowsocks-service/src/local/tun/mod.rs:18:10
   |
18 |     io::{AsyncReadExt, AsyncWriteExt},
   |          ^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `AsyncWriteExt`
unused import: `AsyncWriteExt`

  --> crates/shadowsocks-service/src/local/tun/mod.rs:18:24
   |
18 |     io::{AsyncReadExt, AsyncWriteExt},
   |                        ^^^^^^^^^^^^^

warning: unused import: `Device`
unused import: `Device`

  --> crates/shadowsocks-service/src/local/tun/mod.rs:21:59
   |
21 | use tun::{AsyncDevice, Configuration as TunConfiguration, Device, Layer};
   |                                                           ^^^^^^

Some errors have detailed explanations: E0425, E0432, E0599.
For more information about an error, try `rustc --explain E0425`.
warning: `shadowsocks-service` (lib) generated 3 warnings
`shadowsocks-service` (lib) generated 3 warnings

error: could not compile `shadowsocks-service` due to 5 previous errors; 3 warnings emitted

could not compile `shadowsocks-service` due to 5 previous errors; 3 warnings emitted

> Task :core:cargoBuildArm64 FAILED

Execution failed for task ':core:cargoBuildArm64'.
> Process 'command 'cargo'' finished with non-zero exit value 101

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

```

**Comment by zonyitoo** [comment]
Did you change the `tun` dependency in `Cargo.toml`?

**Comment by dev4u** [comment]
> Did you change the `tun` dependency in `Cargo.toml`?

我没有改动`Cargo.toml`，但是有修改`shadowsocks-android/core/build.gradle.kts`：
```
featureSpec.noDefaultBut(arrayOf(
        "multi-threaded",
        "stream-cipher",
        "aead-cipher-extra",
        "logging",
        "local-flow-stat",
        "local-dns",
        "local-tun",
        "jemalloc"))
```

**Comment by zonyitoo** [comment]
https://github.com/meh/rust-tun/blob/19906c006512acf805f392e744e38aec8adc6618/src/lib.rs#L30-L39

Very interesting.

**Comment by dev4u** [comment]
> https://github.com/meh/rust-tun/blob/19906c006512acf805f392e744e38aec8adc6618/src/lib.rs#L30-L39
> 
> Very interesting.

看来现阶段，不能运行在安卓下……

**Comment by zonyitoo** [comment]
It should work on Android. Since the original crate didn't release a new version that supports tokio v1.x, I should make my `AsyncDevice` manually without its support.

ref https://github.com/shadowsocks/shadowsocks-rust/pull/586

**Comment by dev4u** [comment]
> It should work on Android. Since the original crate didn't release a new version that supports tokio v1.x, I should make my `AsyncDevice` manually without its support.
> 
> ref https://github.com/meh/rust-tun/issues/29

嗯嗯，我等你的消息。谢谢！

**Comment by zonyitoo** [comment]
It is now confirmed working on macOS. Would you like to test if it is working on shadowsocks-android? @madeye @Mygod

**Comment by dev4u** [comment]
> It is now confirmed working on macOS. Would you like to test if it is working on shadowsocks-android? @madeye @Mygod 

现在能编译通过了吗？

**Comment by dev4u** [comment]
```
   Compiling tun v0.5.1
error[E0432]: unresolved import `crate::platform::create`
  --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/lib.rs:28:9
   |
28 | pub use crate::platform::create;
   |         ^^^^^^^^^^^^^^^^^^^^^^^ no `create` in `platform`

error[E0412]: cannot find type `Configuration` in module `platform`
  --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/configuration.rs:38:36
   |
38 |     pub(crate) platform: platform::Configuration,
   |                                    ^^^^^^^^^^^^^ not found in `platform`
   |
help: consider importing this struct
   |
15 | use crate::Configuration;
   |

error[E0412]: cannot find type `Configuration` in module `platform`
  --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/configuration.rs:55:34
   |
55 |         F: FnOnce(&mut platform::Configuration),
   |                                  ^^^^^^^^^^^^^ not found in `platform`
   |
help: consider importing this struct
   |
15 | use crate::Configuration;
   |

error[E0425]: cannot find value `AF_INET` in this scope
   --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/platform/posix/sockaddr.rs:42:31
    |
42  |         if value.sa_family != AF_INET {
    |                               ^^^^^^^
    |
   ::: /.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.98/src/unix/linux_like/mod.rs:679:1
    |
679 | pub const AF_INET: ::c_int = 2;
    | ------------------------------- similarly named constant `_AF_INET` defined here
    |
help: a constant with a similar name exists
    |
42  |         if value.sa_family != _AF_INET {
    |                               ^^^^^^^^
help: consider importing this constant
    |
15  | use libc::AF_INET;
    |

error[E0425]: cannot find value `AF_INET` in this scope
   --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/platform/posix/sockaddr.rs:65:27
    |
65  |         addr.sin_family = AF_INET;
    |                           ^^^^^^^
    |
   ::: /.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.98/src/unix/linux_like/mod.rs:679:1
    |
679 | pub const AF_INET: ::c_int = 2;
    | ------------------------------- similarly named constant `_AF_INET` defined here
    |
help: a constant with a similar name exists
    |
65  |         addr.sin_family = _AF_INET;
    |                           ^^^^^^^^
help: consider importing this constant
    |
15  | use libc::AF_INET;
    |

error[E0412]: cannot find type `c_uint` in this scope
  --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/platform/posix/sockaddr.rs:68:35
   |
68 |             s_addr: ((parts[3] as c_uint) << 24)
   |                                   ^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
15 | use libc::c_uint;
   |
15 | use std::os::raw::c_uint;
   |

error[E0412]: cannot find type `c_uint` in this scope
  --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/platform/posix/sockaddr.rs:69:33
   |
69 |                 | ((parts[2] as c_uint) << 16)
   |                                 ^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
15 | use libc::c_uint;
   |
15 | use std::os::raw::c_uint;
   |

error[E0412]: cannot find type `c_uint` in this scope
  --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/platform/posix/sockaddr.rs:70:33
   |
70 |                 | ((parts[1] as c_uint) << 8)
   |                                 ^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
15 | use libc::c_uint;
   |
15 | use std::os::raw::c_uint;
   |

error[E0412]: cannot find type `c_uint` in this scope
  --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/platform/posix/sockaddr.rs:71:32
   |
71 |                 | (parts[0] as c_uint),
   |                                ^^^^^^ not found in this scope
   |
help: consider importing one of these items
   |
15 | use libc::c_uint;
   |
15 | use std::os::raw::c_uint;
   |

error[E0283]: type annotations needed
   --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/configuration.rs:38:5
    |
35  | #[derive(Clone, Default, Debug)]
    |                 ------- in this derive macro expansion
...
38  |     pub(crate) platform: platform::Configuration,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
    |
    = note: cannot satisfy `_: Default`
note: required by `std::default::Default::default`
   --> /.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/default.rs:116:5
    |
116 |     fn default() -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0283, E0412, E0425, E0432.
For more information about an error, try `rustc --explain E0283`.
error: could not compile `tun` due to 10 previous errors
could not compile `tun` due to 10 previous errors
```

**Comment by zonyitoo** [comment]
> --> /.cargo/registry/src/github.com-1ecc6299db9ec823/tun-0.5.1/src/lib.rs:28:9

Could you confirm that you have upgraded to the latest commit? The latest commit is now using the `tun` on crates.io.

**Comment by zonyitoo** [comment]
Ok. I can reproduce that with Docker build.

**Comment by zonyitoo** [comment]
Compilation finished on Docker. Please help and test on Android.

**Comment by Mygod** [comment]
Cool! I'll take a look sometime.

**Comment by Mygod** [comment]
Where do I start by the way? :)

**Comment by dev4u** [comment]
感谢zonyitoo的快速响应，我也把代码合并到本地的shadowsock-android中做测试。

虽已成功编译，但是当前还不能正常工作。当中也许涉及有vpn service的代码修改：打开连接后，科学网络不通，tg连不上……

我猜测，sslocal建立的tun，与ss android建立的vpn都会接管所有的流量，他们在科学的时候，请求会先流经哪个？后流经的那个是做放行？还是会做什么操作？恐怕需要梳理一遍。

**Comment by zonyitoo** [comment]
Start `sslocal` with `—tun-device-fd-from-path`, then it will listen to this path and get the fd from `VpnService`.

There is another requirement that the system must have a route setting that routes packets with destination to all IPs in the subnet to the

**Comment by dev4u** [comment]
> Where do I start by the way? :)

- 合并ss rust的tun分支
- 修改core的build.gradle.kt，开启local-tun
- 修改启动sslocal的参数，增加: 
`--protocol tun --outbound-bind-interface lo0 --tun-interface-address 10.255.0.1/24`

**Comment by zonyitoo** [comment]
On Android, the `sslocal` should get the device’s fd from VpnService instead of creating a new one.

**Comment by zonyitoo** [comment]
I don’t know if Android could modify the route table in the Application. 

This route setting is required for TCP tunnel because in this implementation it requires this route setting to redirects packets from `tun` to a `TcpListener` that listens to the IP of the `tun` interface’s address.

**Comment by dev4u** [comment]
> I don’t know if Android could modify the route table in the Application. 
> 
> This route setting is required for TCP tunnel because in this implementation it requires this route setting to redirects packets from `tun` to a `TcpListener` that listens to the IP of the `tun` interface’s address.

我印象不是很深，之前我浏览安卓那边的实现，好像在创建vpn对象时，是有设定路由规则。

**Comment by iamwwc** [comment]
I have read all of codes about tun https://github.com/shadowsocks/shadowsocks-rust/blob/4287c2aab8f1826446b68888f58462b4342a0a53/crates/shadowsocks-service/src/local/tun/tcp.rs#L97

I found that you are implement a simple TCP protocol manually, as https://github.com/shadowsocks/shadowsocks-rust/issues/581#issuecomment-890502258 said, Have you consider ordering/retransmission into the situation ?

**Comment by zonyitoo** [comment]
All TCP connections starts with a `SYN`. If retransmission happens, then it will just allocate a new source address and become a completely new connection. I don't think this is an issue here.

**Comment by iamwwc** [comment]
What about packet reordering? If it occurred, local server will send wrong ordering packets to server

**Comment by zonyitoo** [comment]
Packets are handled by a TcpListener (socket), which is a system provided TCP stack. So there is nothing have to be done in the Tun server.

**Comment by zonyitoo** [comment]
Correct.

**Comment by iamwwc** [comment]
On Linux, should we set a default route to routing all packets go through tun device manually? 

I don't see `routing on linux`  in README. only having a route setting related to MacOS

> (OPTIONAL) macOS requires adding a route entry to redirect packets that destinated to `10.155.0.1` (the address of the Tun interface) to the Tun interface itself.

**Comment by zonyitoo** [comment]
Because I have tried on a Linux server, packets destinated to the tun's subnet would be routed properly.

**Comment by iamwwc** [comment]
> Because I have tried on a Linux server, packets destinated to the tun's subnet would be routed properly.

I'm confused. If we don't set route table to routed all TCP/DNS packet to tun device, how we hijack TCP/DNS to give it a fake IP?

**Comment by zonyitoo** [comment]
If you want to route all packets to the `tun` device, you will have to set the `default` route strategy.

The route setting mentioned in the README is for routing faked packets back to the `TcpListener`. It is the default behavior on Linux, so we don't need to do anything.

**Comment by iamwwc** [comment]
> ref #581
> 
> ## Key design
> ### TCP
> image
> 
> 1. Create a `TcpListener` listening on an address in the subnet of `tun` device.
> 2. When `tun` receives TCP packets, it will redirect it to `TcpListener` by modifying `saddr` to a unique fake address and `daddr` to the address that `TcpListener` was listening.
> 3. `TcpListener` accepts a new connection. The connection's `peer_addr` will be the unique fake address. So we can get the original destination address by looking to the `tun`'s address allocation map.
> 
> ### UDP
> The same as the `redir` design, except it sends and receives packets from the `tun` interface.
> 
> ### Possible Applications
> * shadowsocks-android
> * shadowsocks-ios
> * macOS & Linux

But as above picture showed, we should set default route to route all client requests to tun device, so that we can assign a fake ip to destination.

I think tun protocol used for transparent proxy, which you don't set proxy settings manually in every application. Just set a default route, let tun device to handle packet.

> If you want to route all packets to the tun device, you will have to set the default route strategy.

If we don't set the default route, what the tun protocol used for?

**Comment by zonyitoo** [comment]
Well, that's your choice. Setting the `default` route is not the only way to use `tun`.

For example, if you already have a tun device that controls all your packets, then you can simply set sslocal's tun as the outbound's interface.

**Comment by iamwwc** [comment]
I guess I understand.

It like this, right?
```
          outgoing traffic
Surge tun ====> ss-tun
          SO_BINDTODEVICE

          Fake ip
Surge tun ====> ss-tun
```

**Comment by zonyitoo** [comment]
```
ss-tun === fakeip ===> ss-tun
```

File: crates/shadowsocks-service/src/local/tun/tcp.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        } = *(self.translator.lock().await);
+
+        let (conn, is_reply) = if tcp_header.syn && !tcp_header.ack {
+            // 1st SYN, creating a new connection
+            // Allocate a `saddr` for it
+            let saddr = loop {
+                let addr_idx = rand::random::<usize>() % self.free_addrs.len();
+                let port = rand::random::<u16>() % (65535 - 1024) + 1024;
+
+                let addr = SocketAddr::new(self.free_addrs[addr_idx], port);
```
Comment (iamwwc) [inline crates/shadowsocks-service/src/local/tun/tcp.rs:104]: Why we use random up to ten saddrs to determinate which saddr to used? I think use fixed one saddr is enough

**Review by iamwwc — COMMENTED** [review]
Why we should pre allocate ten address for NAT?

File: crates/shadowsocks-service/src/local/tun/tcp.rs (hunk 1 above)
Comment (zonyitoo) [reply crates/shadowsocks-service/src/local/tun/tcp.rs:104]: Well, then you can have at most `(65535 - 1024) * 10` free addresses.

**Comment by iamwwc** [comment]
Does there will be having the windows tun support?

**Comment by zonyitoo** [comment]
PRs are welcome. Since I don't have any Windows device, and there is no official tun interface support in Windows, I won't implement that.

**Comment by iamwwc** [comment]
Will do after finish reading `shadowsocks-rust` codes. And I think the wintun project is a good choice

**Comment by zonyitoo** [comment]
https://crates.io/crates/wintun

You may try to contribute it to https://crates.io/crates/tun .

File: crates/shadowsocks-service/src/local/tun/tcp.rs (hunk 1 above)
Comment (zonyitoo) [reply crates/shadowsocks-service/src/local/tun/tcp.rs:104]: Now the `tun`'s address is included in the free addresses.

**Comment by cirias** [comment]
Hi @dev4u Did you ever make this working on Android?

**Comment by dev4u** [comment]
> Hi @dev4u Did you ever make this working on Android?

考虑到在使用体验上，tun与现在的vpn模式差别不大，所以我没有尝试在手机上tun。
如果你有兴趣，可以试试并欢迎分享你的结果。

**Pull request merged**